File: src/main/java/com/topface/topface/banners/BannerInjector.java
package com.topface.topface.banners;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;

import com.topface.topface.App;
import com.topface.topface.banners.ad_providers.AdProvidersFactory;
import com.topface.topface.banners.ad_providers.IAdsProvider;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created by kirussell on 11/01/15.
 * Controller over ads' sdks
 * Inject banner into provided {@link com.topface.topface.banners.IPageWithAds} with
 * {@link #injectBanner(IPageWithAds)} according to page information
 * </p>
 * When page with banner is destroyed call {@link #cleanUp()} to remove banner and clean container
 * (for example, in onDestroy method of Activity or Fragment)
 */
class BannerInjector implements IBannerInjector {

    private final AdProvidersFactory mProvidersFactory;
    private final List<WeakReference<IPageWithAds>> mUsedPages = new ArrayList<>();
    private Context mContext;

    public BannerInjector(AdProvidersFactory providersFactory, Context context) {
        mProvidersFactory = providersFactory;
        mContext = context;
    }

    @Override
    public void injectBanner(final IPageWithAds page) {
        if (canInject(page)) {
            String banner = lookupBannerFor(page);
            IAdsProvider provider = getProvider(banner);
            showAd(page, provider);
        }
    }

    private boolean canInject(IPageWithAds page) {
        if (!App.from(mContext).getProfile().showAd || page == null) {
            return false;
        }
        PageInfo.PageName pageId = page.getPageName();
        String pageName = pageId.getName();
        Map<String, PageInfo> pagesInfo = App.from(mContext).getOptions().getPagesInfo();
        if (pagesInfo.containsKey(pageName)) {
            String floatType = pagesInfo.get(pageName).floatType;
            if (floatType.equals(PageInfo.FLOAT_TYPE_BANNER)) {
                return true;
            }
        }
        return false;
    }

    private void showAd(final IPageWithAds page, IAdsProvider provider) {
        showAd(page, provider, false);
    }

    private void showAd(final IPageWithAds page, IAdsProvider provider, final boolean isFallbackAd) {
        if (provider != null) {
            final boolean injectInitiated = provider.injectBanner(page,
                    new IAdsProvider.IAdProviderCallbacks() {
                        @Override
                        public void onAdLoadSuccess(View adView) {
                            mUsedPages.add(new WeakReference<>(page));
                        }

                        @Override
                        public void onFailedToLoadAd() {
                            cleanUp(page);
                            if (!isFallbackAd) {
                                injectGag(page);
                            }
                        }
                    });
            if (!injectInitiated && !isFallbackAd) {
                injectGag(page);
            }
        }
    }

    private void injectGag(IPageWithAds page) {
        showAd(page, getProvider(App.from(mContext).getOptions().fallbackTypeBanner), true);
    }

    private IAdsProvider getProvider(String banner) {
        return mProvidersFactory != null ? mProvidersFactory.createProvider(banner, App.from(mContext).getOptions()) : null;
    }

    private String lookupBannerFor(IPageWithAds page) {
        Map<String, PageInfo> pagesInfo = App.from(mContext).getOptions() != null ? App.from(mContext).getOptions().getPagesInfo() : null;
        String pageName = page.getPageName().getName();
        if (pagesInfo != null && pagesInfo.containsKey(pageName)) {
            PageInfo pageInfo = pagesInfo.get(pageName);
            if (pageInfo != null) {
                return pageInfo.getBanner();
            }
        }
        return null;
    }

    public void cleanUp() {
        for (WeakReference<IPageWithAds> page : mUsedPages) {
            IPageWithAds pageWithAds = page.get();
            if (pageWithAds != null) {
                cleanUp(pageWithAds);
            }
        }
    }

    private void cleanUp(IPageWithAds page) {
        ViewGroup container = page.getContainerForAd();
        if (container != null) {
            unbindDrawables(container);
            container.removeAllViews();
        }
    }

    private void unbindDrawables(View view) {
        if (view != null) {
            if (view.getBackground() != null) {
                view.getBackground().setCallback(null);
            }
            if (view instanceof ViewGroup) {
                for (int i = 0; i < ((ViewGroup) view).getChildCount(); i++) {
                    unbindDrawables(((ViewGroup) view).getChildAt(i));
                }
                ((ViewGroup) view).removeAllViews();
            }
        }
    }
}

File: src/main/java/com/topface/topface/banners/BannerInjector.kt
package com.topface.topface.banners

import android.view.View
import com.topface.topface.App
import com.topface.topface.statistics.AdStatistics

/**
 * Created by ppavlik on 31.05.17.
 * Inject banner to container
 */
class BannerInjector : IBannerInjector {

    private val mProvidersFactory by lazy {
        AdProvidersFactory()
    }

    private var mCurrentAdsProvider: IAdsProvider? = null
    private var mPage: IBannerAds? = null
    private var mBannerName = AdProvidersFactory.Companion.BANNER_NONE

    override fun injectBanner(bannerName: String, provider: IAdsProvider, container: IBannerAds) {
        mPage = container
        mBannerName = bannerName
        mCurrentAdsProvider = provider
        showAd(provider)
    }

    override fun cleanUp() =
            mPage?.let { page ->
                page.getContainerForAd()?.let {
                    mCurrentAdsProvider?.clean(page)
                    unbindDrawables(it)
                    it.removeAllViews()
                }
            } ?: Unit


    private fun showAd(provider: IAdsProvider) = showAd(provider, false)

    private fun showAd(provider: IAdsProvider?, isFallbackAd: Boolean) {
        provider?.let { provider ->
            mPage?.let {
                val injectInitiated = provider.injectBanner(it,
                        object : IAdsProvider.IAdProviderCallbacks {
                            override fun onAdLoadSuccess(adView: View) {}

                            override fun onFailedToLoadAd(codeError: Int?) {
                                AdStatistics.sendBannerFailedToLoad(mBannerName, codeError)
                                cleanUp()
                                if (!isFallbackAd) {
                                    injectGag()
                                }
                            }

                            override fun onAdClick() = AdStatistics.sendBannerClicked(mBannerName)

                            override fun onAdShow() = AdStatistics.sendBannerShown(mBannerName)
                        })
                if (!injectInitiated && !isFallbackAd) {
                    injectGag()
                }
            }
        }
    }

    private fun injectGag() {
        mCurrentAdsProvider = mProvidersFactory.createProvider(App.get().options.fallbackTypeBanner)
        showAd(mCurrentAdsProvider, true)
    }

    private fun unbindDrawables(view: View?) {
        view?.background?.callback = null
        if (view is android.view.ViewGroup) {
            for (i in 0..view.childCount - 1) {
                unbindDrawables(view.getChildAt(i))
            }
            view.removeAllViews()
        }
    }
}